### DISCUSSION TITLE
Capture type argument in pattern matching

### DISCUSSION THREAD

**Discussion by FstTesla** [dotnet/csharplang#6070] [Language Ideas]
Is it sometimes needed to cast an expression to a generic type with an unknown type argument (or to check whether an expression fits), in order to call a generic method. At the moment, since there is no language construct that allows this scenario, the only way to achieve this is via reflection with manual "type deconstruction" and "generic method construction".

Let's consider the following code:

```csharp
void Method<T>(Foo<T> foo) { ... }

object expr = ...; // any compatible type is ok, not only object
Method((Foo<???>)expr);
```

This small example can become more complicated in many ways:
- Instead of a direct cast and call, one may want to type-check `expr` and decide what to do with it (something switch-style).  
- `Foo` may have more than 1 type argument.  
- The called method may receive other arguments whose type is `T` or related to it (e.g. `void Method<T>(Foo<T> foo, IEnumerable<T> enumerable) { ... }`).
  

In order to invoke `Method` with `expr`, one must do the following steps:
1. Go up in `expr`'s type hierarchy until `Foo` is found; this procedure differs if the target type is a class or an interface.  
2. Extract the desired type argument(s) and verify whether they match the constraints (if any).  
3. Select the target generic method definition; this can be tricky and cumbersome when the method has overloads.  
4. Use the extracted type arguments to instantiate the constructed method.  
5. Invoke the method.
  

Even with the assumption that the involved symbols are public (i.e. reachable with `nameof` expressions and not subject to sudden change), all the steps above are done via reflection, which means they are *type-unsafe*, *verbose*, *fragile*, *error-prone* and *slow*.

This proposal presents a language construct based on patterns that can be lowered by the compiler to something akin to the steps 1-2 above. Of course, this approach removes the need for steps 3-5, increasing type-safety and reducing the reflection overhead.

```csharp
// EXAMPLE 1 - Basic
class Foo1<T> { ... }
void Method1<T>(Foo1<T> foo) { ... }

if (expr is Foo1<new T> foo) // <- 'new' is the key point
{
    Method1(foo);
}

// EXAMPLE 2 - Class with 2 type arguments
class Foo2<T1, T2> { ... }
void Method2<T1, T2>(Foo2<T1, T2> foo) { ... }

if (expr is Foo2<new T1, new T2> foo)
{
    Method2(foo);
}

// EXAMPLE 3 - Method with constraint
class Foo3<T> { ... }
void Method3<T>(Foo<T> foo) where T : ISomething { ... }

if (expr is Foo3<new T : ISomething> foo)
{
    Method3(foo);
}

// EXAMPLE 4 - Method with "type-related" parameters
class Foo4<T> { ... }
void Method4<T>(Foo<T> foo, IEnumerable<T> enumerable) { ... }

IEnumerable rawEnumerable = ...;
// Note: 'new' only the first time!
if (expr is Foo4<new T> foo && rawEnumerable is IEnumerable<T> enumerableOfT)
{
    Method(foo, enumerableOfT);
}

// EXAMPLE 5 - Basic with discard
class Foo5<T> { ... }
void Method5<T>(Foo5<T> foo) { ... }

// We actually don't care about the type argument itself, we just need 'expr' to be a 'Foo<>'
if (expr is Foo5<new _> foo)
{
    Method5(foo);
}
```

Here are some key observations of the given proposal examples:
- The purpose is adding to the scope a "local" type argument, perhaps with one or more constraint.  
- The construct is indeed an extension of pattern matching expressions, with all the implied expressive power and flexibility.  
- The scope of the introduced type argument(s) is analogous to how pattern matching works w.r.t. local variables.  
- The usage of the keyword `new` is of course just a suggestion and should not be ambiguous from a grammar point of view.  

As the title may suggest, this proposal is inspired by Java's "Capture conversion" (see [JLS §5.1.10](https://docs.oracle.com/javase/specs/jls/se18/html/jls-5.html#jls-5.1.10)), even if its realization in C# (as I imagined it) is of course much different, given both the different way generics are implemented and the peculiarities of the two languages.

Finally, I would like to underline that "use `dynamic`" is not an answer, since the primary purpose of this proposal is compile-time safety, followed by compactness.

**Comment by HaloFour** [comment]
See: #1225

There's really no facility in the runtime to do anything like this.  Type parameters are not like normal variables and you can't treat them as such without reflection.

**Reply by FstTesla** [reply]
Of course. My idea is not to eliminate reflection, but instead to move the burden of "type deconstruction" to the compiler, removing the need of explicit (programmer-side) reflection and reducing the total amount of reflection needed. This brings more type-safety and expressive power.

If you want, I can try to write how I imagine that the examples above would be lowered.
